python_modules/dagster-external/dagster_external/context.py

```diff
@@ -29,44 +29,51 @@
 
 def init_dagster_external() -> "ExternalExecutionContext":
     params = get_external_execution_params()
+    data = _read_input(params)
+    output_stream = _get_output_stream(params)
+
+    context = ExternalExecutionContext(data, output_stream)
+    ExternalExecutionContext.set(context)
+    return context
+
+
+def _read_input(params: ExternalExecutionParams) -> ExternalExecutionContextData:
     if params.input_mode == ExternalExecutionIOMode.stdio:
         with sys.stdin as f:
-            data = json.load(f)
+            return json.load(f)
     elif params.input_mode == ExternalExecutionIOMode.file:
         assert params.input_path, "input_path must be set when input_mode is `file`"
         with open(params.input_path, "r") as f:
-            data = json.load(f)
+            return json.load(f)
     elif params.input_mode == ExternalExecutionIOMode.fifo:
         assert params.input_path, "input_path must be set when input_mode is `fifo`"
         with open(params.input_path, "r") as f:
-            data = json.load(f)
+            return json.load(f)
     elif params.input_mode == ExternalExecutionIOMode.socket:
         assert params.host, "host must be set when input_mode is `socket`"
         assert params.port, "port must be set when input_mode is `socket`"
         with socket.create_connection((params.host, params.port)) as sock:
             sock.makefile("w").write(f"{GET_CONTEXT_MESSAGE}\n")
-            data = json.loads(sock.makefile("r").readline())
-
+            return json.loads(sock.makefile("r").readline())
     else:
         raise Exception(f"Invalid input mode: {params.input_mode}")
 
+
+def _get_output_stream(params: ExternalExecutionParams) -> Union[TextIO, Tuple[str, int]]:
     if params.output_mode == ExternalExecutionIOMode.stdio:
-        output_target = sys.stdout
+        return sys.stdout
     elif params.output_mode == ExternalExecutionIOMode.file:
         assert params.output_path, "output_path must be set when output_mode is `file`"
-        output_target = open(params.output_path, "a")
+        return open(params.output_path, "a")
     elif params.output_mode == ExternalExecutionIOMode.fifo:
         assert params.output_path, "output_path must be set when output_mode is `fifo`"
-        output_target = open(params.output_path, "w")
+        return open(params.output_path, "w")
     elif params.output_mode == ExternalExecutionIOMode.socket:
         assert params.host, "host must be set when output_mode is `socket`"
         assert params.port, "port must be set when output_mode is `socket`"
-        output_target = (params.host, params.port)
+        return (params.host, params.port)
```

Comment:
This is a bit weird and I will change it once I figure out how to make a persistent socket connection work, right now we are creating a new connection for each notification.